src/Silverback.Integration/Messaging/Inbound/ErrorHandling/ErrorPoliciesHelper.cs

```diff
@@ -10,13 +10,32 @@ namespace Silverback.Messaging.Inbound.ErrorHandling
 {
     internal static class ErrorPoliciesHelper
     {
-        public static async Task<bool> ApplyErrorPoliciesAsync(ConsumerPipelineContext context, Exception exception)
+        public static async Task<bool> ApplyErrorPoliciesAsync(
+            ConsumerPipelineContext context,
+            Exception exception)
         {
-            var failedAttempts = context.Consumer.IncrementFailedAttempts(context.Envelope);
+            int failedAttempts = context.Consumer.IncrementFailedAttempts(context.Envelope);
 
             context.Envelope.Headers.AddOrReplace(DefaultMessageHeaders.FailedAttempts, failedAttempts);
+            context.Envelope.Headers.AddOrReplace(DefaultMessageHeaders.InvalidReason, exception.Message);
 
-            var errorPolicyImplementation = context.Envelope.Endpoint.ErrorPolicy.Build(context.ServiceProvider);
+            if (context.Envelope.Headers.Contains(DefaultMessageHeaders.TimestampKey))
+            {
+                context.Envelope.Headers.AddOrReplace(
+                    DefaultMessageHeaders.SourceTimestamp,
+                    context.Envelope.Headers[DefaultMessageHeaders.TimestampKey]);
+                context.Envelope.Headers.Remove(DefaultMessageHeaders.TimestampKey);
+            }
+
+            context.Envelope.Headers.AddOrReplace(
```

Comment:
These two headers only make sense for Kafka, shouldn't be set in `Silverback.Integration` but moved to a specific implementation of the `MoveMessageErrorPolicy` in `Silverback.Integration.Kafka`.
Other implementations of `IBroker` might (and will) use different data for the identifier's key and value (in MQTT or Rabbit you don't even have the concept of offset and partition).

Reply:
According to our discussion, I have now implemented a `KafkaBrokerHeadersEnricher` responsible for setting custom headers for failed messages.